### PR TITLE
Phase 4 Performance Testing: Deployments under load

### PR DESCRIPTION
**Pull request by LJBabbage**
### What is the context of this PR?
This  PR is the write up for deployment of a new version of eQ runner whilst under load.

### How to review 
Make sure it makes sense, and all details that are relevant are captured

### CODE REVIEW

File: doc/performance-testing/cloud-run/0004-stress-test-phase-4/0004_stress-test-phase-4-febuary-2022.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+Although deployment was successful in each test it was not completely seemless. Each test saw a spike in latency of the new instance at the time of switching, this lasted approximately 20 seconds.
+
+The table below shows the amount of requests in that time period and the count of requests breaking particular thresholds.
+
+| Environment | Total requests | Request latency >2s | Request latency >5s | Request latency >10s | Date/Time | Output |
+| --- | --- | --- | --- |--- | --- | --- |
+| eq-stress-test-340815 (Test1) | 45,034 | 6017 | 3094 | 58 | 09-02-2022 20:00 |[output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2022-02-09T20:00:45/) |
+| eq-stress-test-340815 (Test2) | 43,885 | 7540 | 5453 | 771 | 10-02-2022 13:18:38 |[output](https://console.cloud.google.com/storage/browser/eq-stress-injector-07122021-outputs/stress-test/2022-02-10T13:18:38/) |
+| Trail run | 45,059 | 7443 | 3385 | 228 | 09-02-2022 13:14:03 | ephemeral GCP project |
```
Comment (berroar) [inline doc/performance-testing/cloud-run/0004-stress-test-phase-4/0004_stress-test-phase-4-febuary-2022.md:56]:
```suggestion
| Trial run | 45,059 | 7443 | 3385 | 228 | 09-02-2022 13:14:03 | ephemeral GCP project |
```

**Review by berroar — APPROVED** [review]
All looks good to me 👍

File: doc/performance-testing/cloud-run/0004-stress-test-phase-4/0004_stress-test-phase-4-febuary-2022.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+Trial
+
+![](0004_trial-response-time-and-requests.png)
+![](0004_trial-instances.png)
+![](0004_trial-cpu.png)
+
+## Conclusion
+Whether it was the test settings or the timeliness of the instance switching there does seem to be a gap lasting approximately 20 seconds when the new instance has significantly higher latency.
+
+The results of the test showed that 15% of requests at the time were above 2 seconds. In 2 of the tests aftershocks were also seen for the next 3-4 minutes, where a small proportion of requests also took greater than 2 seconds (approximately 50 requests)
```
Comment (petechd) [inline doc/performance-testing/cloud-run/0004-stress-test-phase-4/0004_stress-test-phase-4-febuary-2022.md:82]: I don't want to be that guy.. but full stop missing.

**Review by petechd — APPROVED** [review]

File: doc/performance-testing/cloud-run/0004-stress-test-phase-4/0004_stress-test-phase-4-febuary-2022.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
+# Stress Test Phase 4 February 2022
+
+To test deployment of a new version of eQ runner whilst under load.
+
+The test was conducted 3 times, one trial run in a separate GCP project before 2 tests using eq-stress-test-340815. The trial results have been included to highlight a potential issue with latency on deployment.
+
+In each test, runner version 3.9.4 was upgraded to 3.9.5. The only exception was the last test which was downgraded from 3.9.5 to 3.9.4, this seemed a completely valid scenario and due to no breaking changes in the versions had no ill effect.
```
Comment (pricem14pc) [inline doc/performance-testing/cloud-run/0004-stress-test-phase-4/0004_stress-test-phase-4-febuary-2022.md:7]:
```suggestion
In each test, runner version 3.94.0 was upgraded to 3.95.0 The only exception was the last test which was downgraded from 3.95.0 to 3.94.0, this seemed a completely valid scenario and due to no breaking changes in the versions had no ill effect.
```

**Comment by pricem14pc** [comment]
Rather than the spike being a consequence of the GCP revision replacement (A -> B) we do see a similar pattern following all cold starts. This can be seen in the daily performance tests. My thoughts are that while we need to investigate (a) improving cold starts and (b) improving absorbing of an increase in load (e.g. from 0 requests to hundreds of requests/sec), the lack of errors during the revision replacement is encouraging.

**Review by petechd — APPROVED** [review]

**Review by pricem14pc — APPROVED** [review]

**Pull request merged**